server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java

```diff
@@ -222,6 +222,31 @@ public T actionGet(long timeout, TimeUnit unit) {
         }
     }
 
+    /**
+     * Return the result of this future, if it has been completed successfully, or throw the exception with which it was completed
+     * exceptionally. It is not valid to call this method if the future is incomplete.
+     */
+    public T result() {
+        if (isDone()) {
+            return FutureUtils.get(this, 0L, TimeUnit.NANOSECONDS);
```

Comment:
Given that `result` is non-blocking, it feels odd that it can throw `InterruptedException` (though wrapped). I suppose we do not use interrupt (except in tests), so it is kind of a non-issue.

But could we not instead return `sync.getValue()` here? And perhaps let `result()` still throw `ExecutionException` to let it have same semantics as `get()`. Clients wanting to avoid that can use `actionResult` instead.

I would find that slightly nicer.

Reply:
One of my motivations for this change was to be able to clean up the way that these things expose exceptions (and indeed getting rid of the spurious `InterruptedException` entirely) although I was going to do that in a follow-up. I can do that here instead if you'd prefer.

It's not quite true that we avoid interrupts in production code: the bad handling of interrupts in this area today is the reason for needing to use a `CompletableFuture` in `RecoverySourceHandler` instead of one of our own futures.

Reply:
I pushed 6e2e48a759adc92b4e68deb95f3076de7b3a5b9d - I thought it would be a little more invasive than this hence deserving of a separate PR but on reflection this isn't a big change.

Reply:
Thanks for doing this. I notice that `result()` still handles exceptions differently from `get()`, is there a good reason that you want to avoid the `ExecutionException` for `result()`? I find it odd to have this subtle difference, unless there is a good reason for it?

Reply:
I intend `X.result()` to be a replacement for `FutureUtils.get(X, 0L, TimeUnit.Y)`, not `X.get()`. There's only one place we invoke `get()` directly in production, in `CompositeIndexEventListener`, and the exception behaviour doesn't matter here (I probably should have used `FutureUtils#get` when I introduced this in #92979).

IMO it's a bit less useful to make `result()` declare a checked exception like `ExecutionException`, although I could do this and introduce another `FutureUtils` wrapper to mangle it back to a `RuntimeException` if you'd prefer.